UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,696 @@
+        } else if (AsciiStrCmp (TempStr, "format") == 0) {
+          TempStr = (CHAR8 *)(PropertyPtr->Data);
+          if (AsciiStrCmp (TempStr, "a8r8g8b8") == 0) {
+            GraphicsInfo->GraphicsMode.PixelFormat = PixelRedGreenBlueReserved8BitPerColor;
+          } else if (AsciiStrCmp (TempStr, "a8b8g8r8") == 0) {
+            GraphicsInfo->GraphicsMode.PixelFormat = PixelBlueGreenRedReserved8BitPerColor;
+          } else if (AsciiStrCmp (TempStr, "a16b16g16r16") == 0) {
+            GraphicsInfo->GraphicsMode.PixelFormat = PixelFormatMax;
+          }
+        }
+      }
+    } else if (AsciiStrCmp (NodePtr->Name, "options") == 0) {
+      DEBUG ((DEBUG_INFO, "  Found options node (%08X)", Node));
+
+      for (SubNode = FdtFirstSubnode (Fdt, Node); SubNode >= 0; SubNode = FdtNextSubnode (Fdt, SubNode)) {
+        NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + SubNode + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+        DEBUG ((DEBUG_INFO, "\n      SubNode(%08X)  %a", SubNode, NodePtr->Name));
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-images@", AsciiStrLen ("upl-images@")) == 0) {
+          DEBUG ((DEBUG_INFO, "  Found image@ node \n"));
+          //
+          // Build PayloadBase HOB .
+          //
+          PayloadBase = BuildGuidHob (&gUniversalPayloadBaseGuid, sizeof (UNIVERSAL_PAYLOAD_BASE));
+          ASSERT (PayloadBase != NULL);
+          if (PayloadBase == NULL) {
+            return EFI_OUT_OF_RESOURCES;
+          }
+
+          PayloadBase->Header.Revision = UNIVERSAL_PAYLOAD_BASE_REVISION;
+          PayloadBase->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_BASE);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr", &TempLen);
+
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data64       = (UINT64 *)(PropertyPtr->Data);
+            StartAddress = Fdt64ToCpu (*Data64);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  entry"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+            PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-params", AsciiStrLen ("upl-params")) == 0) {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr_width", &TempLen);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+            BuildCpuHob (SizeOfMemorySpace, PcdGet8 (SizeOfIoSpace));
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "pci-enum-done", &TempLen);
+          if (TempLen > 0) {
+            Data32      = (UINT32 *)(PropertyPtr->Data);
+            PciEnumDone = (UINT8)Fdt32ToCpu (*Data32);
+            DEBUG ((DEBUG_INFO, "  Found PciEnumDone (%08X)\n", PciEnumDone));

```

Comment:
PciEnumDone should be assigned to UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES->ResourceAssigned, in our case both should be FALSE as we rely on Payload to do PciEnumeration.

Reply:
resolved